components/provider_ppucd/src/parse_ppucd.rs

```diff
@@ -48,33 +52,102 @@ fn is_property_line(line: &str) -> bool {
-        // TODO: ask Markus what to do with the property lines that appear to have
-        // no canonical name
-        // property;Binary;;alnum
-        // property;Binary;;blank
-        // property;Binary;;graph
-        // property;Binary;;print
-        // property;Binary;;xdigit
-        if line_parts[0].is_empty() {
-            line_parts.drain(0..1);
-        }
+        update_aliases(binary_prop_aliases, line);
+    } else if prop_type == &"Enumerated" {
+        update_aliases(enum_prop_aliases, line);
+    }
+}
 
-        let canonical_name = line_parts[0];
-        let all_names: Vec<&'s str> = line_parts.iter().copied().collect();
-        let all_names_set: HashSet<&'s str> = all_names.into_iter().collect();
-        prop_aliases.insert(canonical_name, all_names_set);
+/// Parse enum property value definition line.
+fn update_enum_val_aliases<'s>(
+    enum_val_aliases: &mut HashMap<&'s str, HashMap<&'s str, HashSet<&'s str>>>,
+    line: &'s str,
+) {
+    let line_parts = split_line(&line);
+    let mut line_parts: &[&str] = line_parts.as_slice();
+    assert_eq!(&"value", &line_parts[0]);
+    line_parts = &line_parts[1..];
+    let enum_prop_name = line_parts[0];
+    let enum_prop_val = line_parts[1];
+    enum_val_aliases
+        .entry(enum_prop_name)
+        .or_insert_with(HashMap::new);
+    let enum_val_alias_map: &mut HashMap<&str, HashSet<&str>> =
+        enum_val_aliases.get_mut(&enum_prop_name).unwrap();
+    enum_val_alias_map
+        .entry(enum_prop_val)
+        .or_insert_with(HashSet::new);
+    let enum_prop_val_aliases: &mut HashSet<&str> =
+        enum_val_alias_map.get_mut(&enum_prop_val).unwrap();
+    enum_prop_val_aliases.insert(enum_prop_val);
+    line_parts = &line_parts[2..];
+    // What remains of line_parts are all of the remaining aliases for this
+    // enumerated property's value
+    for alias in line_parts {
+        enum_prop_val_aliases.insert(alias);
     }
 }
 
+/// Return a new map in which any binary exclusion values
+/// (in other words, values for binary properties that are prefixed with a
+/// minus sign, as described in the PPUCD documentation, ex: "-Gr_Base")
+/// are not included and any existing include values for the binary property
+/// (ex: "Gr_Base") are accordingly also removed from the map.
+fn apply_exclude_vals_for_binary_props<'s>(
+    prop_vals: HashMap<&'s str, &'s str>,
+) -> HashMap<&'s str, &'s str> {
+    let mut prop_names: HashSet<&'s str> = prop_vals.keys().copied().collect();
+    // If we see "-Gr_Base", then remove both "Gr_Base" and "-Gr_Base".
+    for prop_name in prop_vals.keys() {
+        if let Some(unadorned_prop_name) = prop_name.strip_prefix('-') {
+            let orig_prop_name: &str = unadorned_prop_name;
+            prop_names.remove(&orig_prop_name);
+            prop_names.remove(prop_name);
+        }
+    }
+    prop_vals
+        .clone()
+        .into_iter()
+        .filter(|(k, _)| prop_names.contains(k))
+        .collect()
```

Comment:
Drive-by: 

It looks like you've got some unnecessary allocations here.

The first is the `.clone()`. Since the map is being moved in and then moved out, cloning the map before calling `into_iter()` will construct a new map, then move the new map into the iterator, when we could just move the existing map into the iterator. 

In general, re-collecting into a new map by moving into an iterator may also cause unnecessary allocations (unless the compiler is smart enough to see what's going on an optimize it to use the existing map anyway, which would be awesome and impressive).

I recommend using the `retain()` method, which will just filter the keys within the existing hash map.

```rust
fn apply_exclude_vals_for_binary_props<'s>(
    mut prop_vals: HashMap<&'s str, &'s str>,
) -> HashMap<&'s str, &'s str> {
    // ...
    // HashSet stuff here
    // ...
    prop_vals.retain(|prop_name, _| {
        prop_names.contains(prop_name)
    });
    prop_vals
}
```

Note, you'll have to make the `prop_vals` argument `mut` to use `retain()` on it.

Reply:
I tried this change out, but then I got the "[Cannot borrow `x` as mutable more than once at a time](https://stackoverflow.com/questions/31281155/cannot-borrow-x-as-mutable-more-than-once-at-a-time)" type of error at the call sites for this function.  The code there in `get_code_point_prop_vals` looks like
```
...
    let mut prop_vals: HashMap<&'s str, &'s str> = HashMap::new();
    prop_vals.clone_from(defaults);
    for (range, block_prop_vals) in blocks {
        if range.contains_u32(code_point) {
            prop_vals.extend(block_prop_vals);
            prop_vals = apply_exclude_vals_for_binary_props(prop_vals);
        }
    }
```
and the compiler error complains about the 2 references to `prop_vals` being in the same lexical scope.

Is there a way to work around this? Also, I should have mentioned this issue in the followup issue #454 for improving performance, so if you have ideas, feel free to add them there.

Reply:
I take back what I said -- I was getting confused by the compiler's suggestions and to which side of the colon the `mut` belongs not knowing the implications of the difference. Your suggested change did indeed work.

Btw, I made an extra modification on top of that -- I changed the signature from taking `mut prop_vals: HashMap<&'s str, &'s str>` with a return type of `HashMap<&'s str, &'s str>` to having no return value (since it's now just mutating the map) and taking `prop_vals: &mut HashMap<&'s str, &'s str>`.

Reply:
Sorry @echeran I was out the past two days. 

Glad you got this working. 

Re: your extra modification,

Totally! I think it's probably more typical to just take a `&mut HashMap`, but sometimes the move-in, move-out semantics are more desirable. I didn't look  further into the context of its usage to form an opinion about it.